plugin/src/main/java/com/google/errorprone/xplat/checker/LazyInitBan.java

```diff
@@ -0,0 +1,409 @@
+
+          // if the assignments are in the wrong order
+          if (!(((IdentifierTree) var).getName().equals(nonSyncIdent.getSimpleName()) && foundIdents
+              .contains(((IdentifierTree) nestedVar).getName()))) {
+            wrongAssignmentTree.editBox(var, ((IdentifierTree) var).getName(),
+                ((IdentifierTree) nestedVar).getName());
+            return IfTreeReturn.WRONG_ASSIGNMENT_ORDER;
+          }
+          foundAssignment = true;
+        }
+      }
+    }
+
+    // result should be true if an assignment was found and the field is not configured correctly
+    boolean result = foundAssignment && !checkFieldSymbolNonSync(nonSyncIdent);
+
+    if (result) {
+      return IfTreeReturn.MATCH;
+    } else {
+      return IfTreeReturn.NO_MATCH;
+    }
+  }
+
+  /**
+   * Main loop to iterate though the statements in a block and check for matched lazy init patters.
+   *
+   * @param statements  list of statements inside a block
+   * @param foundIdents set of found identifiers
+   * @param methodSync  whether the method is synchronized
+   * @param tree        matched method tree
+   * @return Description that describes the error or non error found
+   */
+  private Description statementLoop(List<? extends StatementTree> statements, Set<Name> foundIdents,
+      boolean methodSync, MethodTree tree, VisitorState state) {
+
+    //true if method is nonsync variety
+    boolean nonSync = false;
+    Symbol nonSyncIdent = null;
+
+    //true if return needs to be checked for the local var instead of the field
+    boolean checkReturn = false;
+
+    //true if the assignment order is wrong
+    boolean wrongAssignmentOrder = false;
+    ExpressionBox wrongAssignmentTree = new ExpressionBox();
+
+    for (StatementTree stmt : statements) {
+      if (stmt.getKind() == Kind.VARIABLE) {
+
+        VariableTree var = ((VariableTree) stmt);
+
+        if (var.getInitializer().getKind() != Kind.IDENTIFIER) {
+          return Description.NO_MATCH;
+        }
+
+        Name name = ((IdentifierTree) var.getInitializer()).getName();
+
+        Symbol fieldSymbol = ASTHelpers.getSymbol(var).enclClass().members().findFirst(
+            (com.sun.tools.javac.util.Name) name);
+
+        if (fieldSymbol.isLocal()) {
+          return Description.NO_MATCH;
+        }
+
+        foundIdents.add(name);
+        foundIdents.add(var.getName());
+        nonSyncIdent = fieldSymbol;
+        nonSync = true;
+
+      } else if (stmt.getKind() == Kind.IF) {
+
+        if (nonSync) {
+          IfTreeReturn result = ifTreeNonSync((IfTree) stmt, foundIdents, nonSyncIdent,
+              wrongAssignmentTree);
+          if (result == IfTreeReturn.NO_MATCH) {
+            checkReturn = true;
+          } else if (result == IfTreeReturn.WRONG_ASSIGNMENT_ORDER) {
+            wrongAssignmentOrder = true;
+            checkReturn = true;
+          }
+
+        } else {
+          if (!ifTreeSync((IfTree) stmt, foundIdents, methodSync)) {
```

Comment:
Makes me wonder if we are in a synchronized block, could we just skip the whole check? Even if `ifTreeSync` returns true, we really just fall through here, right?

Reply:
If `IfTreeSync` returns true, it still goes on to check the return statement, and uses the field it gets from the return to generate an error message which includes the field name